JIT: Clone ADDR(LCL) for DUP

EgorBo opened the pull request:
Low-hanging fruit, closes https://github.com/dotnet/runtime/issues/57055 and codegen issues in this tweet https://twitter.com/badamczewski01/status/1484990815910957058

Diffs: https://dev.azure.com/dnceng/public/_build/results?buildId=1572041&view=ms.vss-build-web.run-extensions-tab

EgorBo committed:
Clone ADDR(LCL) for DUP

tannergooding commented on src/coreclr/jit/importer.cpp:
```diff
@@ -14411,15 +14411,27 @@ void Compiler::impImportBlockCode(BasicBlock* block)
+                bool needsTemp = true;
+                if (opts.OptimizationEnabled())
+                {
+                    if (op1->IsIntegralConst(0) || op1->IsFPZero() || op1->IsLocal())
```
I know it was already the case, but `IsFPZero` here is likely a bug.

It's going to catch both `+0.0` and `-0.0` and only the first is "simple". We should be using `IsFloatPositiveZero` instead which ensures its only `+0.0`.

EgorBo replied:
Thanks! addressed

tannergooding commented on src/coreclr/jit/importer.cpp:
```diff
@@ -14411,15 +14411,27 @@ void Compiler::impImportBlockCode(BasicBlock* block)
+                    {
+                        needsTemp = false;
+                    }
+                    else if (tree->TypeIs(TYP_BYREF) && tree->OperIs(GT_ADDR) && tree->gtGetOp1()->IsLocal())
```
Should this also check for `TYP_I_IMPL` or `GT_IND`? What about the `GT_OBJ` scenario?

Asking since these are all special cases we handle in `impSIMDPopStack` to cover `ref T`, `T*`, and a few other scenarios: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/simd.cpp#L1225-L1305

So I'd imagine they may also qualify here in some circumstances.

SingleAccretion replied:
This is checking for addresses only, so, no need to check for indirections.

This should probably be using `impIsAddressInLocal` to catch both `LCL_VAR_ADDR` and `LCL_FLD_ADDR` - equivalent trees.

EgorBo committed:
Address feedback

EgorBo committed:
Clean up

kunalspathak commented:
coreclr_tests, libraries_tests and libraries.crossgen2 seems to have regression `core_clr` and `libraries_test` the most:

```coreclr_tests.pmi.Linux.arm64.checked.mch:

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 168170904 (overridden on cmd)
Total bytes of diff: 168332188 (overridden on cmd)
Total bytes of delta: 161284 (0.10 % of base)
    diff is a regression.
    relative diff is a regression.

coreclr_tests.pmi.windows.x64.checked.mch:

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 127473217 (overridden on cmd)
Total bytes of diff: 127708677 (overridden on cmd)
Total bytes of delta: 235460 (0.18 % of base)
    diff is a regression.
    relative diff is a regression.

```

![image](https://user-images.githubusercontent.com/12488060/150867865-06dd95b6-b9e5-486b-b6b0-e4b3e65d6a1d.png)

I guess there is some more tuning needed to make sure we don't do unnecessary copies? I will have to see if some of them can be eliminated by https://github.com/dotnet/runtime/pull/64130.

EgorBo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into jit-opt-dup

EgorBo committed:
Fix regressions

EgorBo committed:
Fix regressions

EgorBo commented:
> coreclr_tests, libraries_tests and libraries.crossgen2 seems to have regression core_clr and libraries_test the most:

Thanks, I didn't realize that for compDbg I had to always "clone" rather than making a temp, fixed.

> This should probably be using impIsAddressInLocal to catch both LCL_VAR_ADDR and LCL_FLD_ADDR - equivalent trees.

I am seeing a lot of size regression with `impIsAddressInLocal` so I decided to leave it as is. Also, added a workaround to fix more regressions where clonned byref is used to zero a local struct, e.g. [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAJDSA6AJQFcA7DbAWxiYDCEHgAc8MKAGUJAN2xgYAZwDcVKoQDMDYnQFUA3lTrG6ijFDZgMdALIBPSecsYjJw5ROe6uCBwDmdACGqpSuXnQA2jYwGAAWEAAmAJKiuAAU0XGJKSK4APIi3L6KTAByEEkcuNgcNX4AlAC6YV6aDCh0ACIQAOrQANZp9XT6dAC+LROhHiZRMfHJqRnz2akFRRwl5ZXVtf5NLYxI7XQAKkoYQ3QAvAB8dBwwAO62Dk5WQ0zdfVCD9SFjIA===)

There are still some regressions since struct enreg is not always a good thing e.g. when we need to copy struct

kunalspathak reviewed (commented):
> There are still some regressions since struct enreg is not always a good thing e.g. when we need to copy struct. Overall, diff is negative and the example from the tweet is improved.

Can you paste the diff of regression with a sample example? It will help me in my future work on structs.

EgorBo commented:
@kunalspathak here is a minimal repro:
```csharp
public virtual DateTimeOffset? EndTime { get; set; }

public TimeSpan Duration() => 
    (EndTime?.Subtract(DateTimeOffset.MinValue)).Value;
```
Diff: https://www.diffchecker.com/QZjlpePR

All of them look like this (quite often connected with `Nullable<>`). I feel like I should not be doing this opt for something that won't be enregistrated but it's too early to check I guess (I tried `structPromotionHelper->CanPromoteStructType` but it didn't help much)

kunalspathak reviewed (approved):
LGTM
